Realm/RLMSyncSubscription.mm

```diff
@@ -0,0 +1,483 @@
+        NSError *err = [[NSError alloc] initWithDomain:@"subscription_set" code:-1 userInfo:@{@"reason":@(error.what())}];
+        completionBlock(&err);
+    }
+    _subscriptionSet->get_state_change_notification(realm::sync::SubscriptionSet::State::Complete)
+        .get_async([completionBlock](realm::StatusWith<realm::sync::SubscriptionSet::State> state) mutable noexcept {
+            if (state.is_ok()) {
+                auto value = state.get_value();
+                completionBlock(nil);
+            } else {
+                NSError* error = [[NSError alloc] initWithDomain:@"sync_subscriptions" code:state.get_status().code() userInfo:@{@"reason": @(state.get_status().reason().c_str())}];
+                completionBlock(&error);
+            }
+        });
+}
+
+typedef void(^RLMSyncSubscriptionCallback)(NSError * _Nullable error);
+
+- (BOOL)writeAsync:(__attribute__((noescape)) void(^)(void))block
+          callback:(RLMSyncSubscriptionCallback)callback {
+    [NSException raise:@"NotImplemented" format:@"Needs Implementation"];
+    return NULL;
+}
+
+#pragma mark - Find subscription
+
+- (nullable RLMSyncSubscription *)subscriptionWithName:(NSString *)name {
+    std::string str = std::string([name UTF8String]);
+    auto iterator = _subscriptionSet->find(str);
+    if (iterator != _subscriptionSet->end()) {
+        return [[RLMSyncSubscription alloc] initWithSubscription:*iterator
+                                                 subscriptionSet:self];
+    }
+    return NULL;
+}
+
+- (nullable RLMSyncSubscription *)subscriptionWithClassName:(NSString *)objectClassName
+                                                      where:(NSString *)predicateFormat, ... {
+    va_list args;
+    va_start(args, predicateFormat);
+    return [self subscriptionWithClassName:objectClassName
+                                     where:predicateFormat
+                                      args:args];
+    
+}
+
+- (nullable RLMSyncSubscription *)subscriptionWithClassName:(NSString *)objectClassName
+                                                      where:(NSString *)predicateFormat
+                                                       args:(va_list)args {
+    return [self subscriptionWithClassName:objectClassName
+                                 predicate:[NSPredicate predicateWithFormat:predicateFormat arguments:args]];
+}
+
+- (nullable RLMSyncSubscription *)subscriptionWithClassName:(NSString *)objectClassName
+                                                  predicate:(NSPredicate *)predicate {
+    RLMClassInfo& info = _realm->_info[objectClassName];
+    auto query = RLMPredicateToQuery(predicate, info.rlmObjectSchema, _realm.schema, _realm.group);
+    auto iterator = _subscriptionSet->find(query);
+    if (iterator != _subscriptionSet->end()) {
+        return [[RLMSyncSubscription alloc] initWithSubscription:*iterator
+                                                 subscriptionSet:self];
+    }
+    return NULL;
```

Comment:
```suggestion
    if (auto it = _subscriptionSet->find(query); it != _subscriptionSet->end()) {
        return [[RLMSyncSubscription alloc] initWithSubscription:*it
                                                 subscriptionSet:self];
    }
    return nil;
```

`it` is the idiomatic name for an iterator variable when there's just one and you don't need to distinguish different iterators from each other. When possible the find should be inside the if, and `nil` should be used for obj-c pointers.